storage/sql.go

```diff
@@ -0,0 +1,378 @@
+	}()
+
+	select {
+	case <-ctx.Done():
+		return nil, ErrIteratorDone
+	case err := <-t.errCh:
+		return nil, HandleSQLError(err)
+	case result := <-t.resultCh:
+		return result, nil
+	}
+}
+
+// ToArray converts the tupleIterator to an []*openfgapb.Tuple and a possibly empty continuation token. If the
+// continuation token exists it is the ulid of the last element of the returned array.
+func (t *SQLTupleIterator) ToArray(ctx context.Context, opts PaginationOptions) ([]*openfgapb.Tuple, []byte, error) {
+	var res []*openfgapb.Tuple
+	for i := 0; i < opts.PageSize; i++ {
+		tupleRecord, err := t.next(ctx)
+		if err != nil {
+			if err == ErrIteratorDone {
+				return res, nil, nil
+			}
+			return nil, nil, err
+		}
+		res = append(res, tupleRecord.AsTuple())
+	}
+
+	// Check if we are at the end of the iterator. If we are then we do not need to return a continuation token.
+	// This is why we have LIMIT+1 in the query.
+	tupleRecord, err := t.next(ctx)
+	if err != nil {
+		if errors.Is(err, ErrIteratorDone) {
+			return res, nil, nil
+		}
+		return nil, nil, err
+	}
+
+	contToken, err := json.Marshal(NewContToken(tupleRecord.Ulid, ""))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return res, contToken, nil
+}
+
+func (t *SQLTupleIterator) Next(ctx context.Context) (*openfgapb.Tuple, error) {
+	record, err := t.next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return record.AsTuple(), nil
+}
+
+func (t *SQLTupleIterator) Stop() {
+	t.rows.Close()
+}
+
+type SQLObjectIterator struct {
+	rows     *sql.Rows
+	resultCh chan *openfgapb.Object
+	errCh    chan error
+}
+
+// NewSQLObjectIterator returns a tuple iterator for Postgres
+func NewSQLObjectIterator(rows *sql.Rows) *SQLObjectIterator {
+	return &SQLObjectIterator{
+		rows:     rows,
+		resultCh: make(chan *openfgapb.Object, 1),
+		errCh:    make(chan error, 1),
+	}
+}
+
+var _ ObjectIterator = (*SQLObjectIterator)(nil)
+
+func (o *SQLObjectIterator) Next(ctx context.Context) (*openfgapb.Object, error) {
+	go func() {
+		if !o.rows.Next() {
+			if err := o.rows.Err(); err != nil {
+				o.errCh <- err
+				return
+			}
+			o.errCh <- ErrIteratorDone
+			return
+		}
+
+		var objectID, objectType string
+		if err := o.rows.Scan(&objectType, &objectID); err != nil {
+			o.errCh <- err
+			return
+		}
+
+		if err := o.rows.Err(); err != nil {
+			o.errCh <- err
+			return
+		}
+
+		o.resultCh <- &openfgapb.Object{
+			Type: objectType,
+			Id:   objectID,
+		}
+	}()
+	select {
+	case <-ctx.Done():
+		return nil, ErrIteratorDone
+	case err := <-o.errCh:
+		return nil, HandleSQLError(err)
+	case result := <-o.resultCh:
+		return result, nil
+	}
+}
+
+func (o *SQLObjectIterator) Stop() {
+	_ = o.rows.Close()
+}
+
+func HandleSQLError(err error, args ...interface{}) error {
+	if errors.Is(err, sql.ErrNoRows) {
+		return ErrNotFound
+	} else if errors.Is(err, ErrIteratorDone) {
+		return ErrIteratorDone
+	} else if strings.Contains(err.Error(), "duplicate key value") { // Postgres
```

Comment:
This case and the one below (for `MySQLError`) are specific to those datastores. I'm wondering if we even need this `HandleSQLError`? I think I'd prefer to see this error handling inline in the specific code paths where these conditions to have to take place.

Also it feels weird to me to blindly convert `sql.ErrNoRows` to `ErrNotFound`, because that implies something about the caller's expected behavior. If I were to blindly handle an error by calling `HandleSQLError(err)` and I didn't want to convert `sql.ErrNoRows` to `ErrNotFound`, then this code would be working against me. This is why I believe the inline error handling approach is more obvious.

WDYT?

Reply:
Yes, this seems like a good idea. I'll refactor. Thanks!

Reply:
This is huge and error-prone. Mind if I do it in a follow-up?

Reply:
Sure